environment/environment.go

```diff
@@ -67,64 +66,64 @@ type Environment struct {
 
 	ID       string
 	Name     string
-	Source   string
 	Worktree string
 
 	Services []*Service
 
 	History History
 
+	Notes Notes
+
 	mu        sync.Mutex
 	container *dagger.Container
 }
 
-func (env *Environment) apply(ctx context.Context, name, explanation, output string, newState *dagger.Container) error {
-	if _, err := newState.Sync(ctx); err != nil {
-		return err
-	}
+type Notes []string
 
-	env.mu.Lock()
-	defer env.mu.Unlock()
-	revision := &Revision{
-		Version:     env.History.LatestVersion() + 1,
-		Name:        name,
-		Explanation: explanation,
-		Output:      output,
-		CreatedAt:   time.Now(),
-		container:   newState,
-	}
-	containerID, err := revision.container.ID(ctx)
+func (n *Notes) Add(format string, a ...any) {
+	*n = append(*n, fmt.Sprintf(format, a...))
+}
+
+func (n *Notes) Clear() {
+	*n = []string{}
+}
+
+func (n *Notes) String() string {
+	return strings.Join(*n, "\n")
+}
+
+func (env *Environment) Export(ctx context.Context) (rerr error) {
+	_, err := env.container.Directory(env.Config.Workdir).Export(
+		ctx,
+		env.Worktree,
```

Comment:
this is a bit of a hole in the abstraction. not a problem for now but a good thing to noodle on.

Reply:
Agreed.

I was thinking it's not *that* bad because the interface of environments is pretty straightforward: **worktree in** (`New`, `Load`), **worktree out** (`Export`). The rest is out of scope for the environment.

Reply:
the hard part, and a big part of why #59 felt like boiling the ocean, is eventually it'd great to point these at a network remote and these lil abstraction violations will prevent us from doing that.

Reply:
Yeah. I actually kept #59 in mind while doing this -- I was kinda hoping that this would be compatible. 1) Keep the environment dumb (load stuff <here>, dump stuff <there>). 2) All that logic moved as-is (for now) in a layer above 3) Eventually `Repository` could use the `Remote` abstraction (e.g. rather than integrating `Remote` in all of `Environment`, we could integrate it into `Repository` only?)

Just realized that maybe I'm missing something. I assumed we'd still use local storage as a staging area, even for other storage implementation, but I see now you meant to be able to avoid doing that and going straight to the alternative storage without stopping by local disk?